File: ruby/strain/strain.rb
module Strain
  refine Array do
    def keep
      [].tap do |array|
        each { |item| array << item if yield item }
      end
    end

    def discard
      [].tap do |array|
        each { |item| array << item unless yield item }
      end
    end
  end
end

File: haskell/strain/src/Strain.hs
module Strain
  ( keep
  , discard
  )
where

keep :: (a -> Bool) -> [a] -> [a]
keep _ [] = []
keep fn (x : xs) | fn x      = x : keep fn xs
                 | otherwise = keep fn xs


discard :: (a -> Bool) -> [a] -> [a]
discard fn = keep $ not . fn
